src/SignalR/common/Shared/SystemTextJsonExtensions.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.Text;
+using System.Text.Json;
+
+namespace Microsoft.AspNetCore.Internal
+{
+    internal static class SystemTextJsonExtensions
+    {
+        public static bool CheckRead(this ref Utf8JsonReader reader)
+        {
+            if (!reader.Read())
+            {
+                throw new InvalidDataException("Unexpected end when reading JSON.");
+            }
+
+            return true;
+        }
+
+        public static void EnsureObjectStart(this ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new InvalidDataException($"Unexpected JSON Token Type '{GetTokenString(reader.TokenType)}'. Expected a JSON Object.");
+            }
+        }
+
+        public static string GetTokenString(JsonTokenType tokenType)
+        {
+            switch (tokenType)
+            {
+                case JsonTokenType.None:
+                    break;
+                case JsonTokenType.StartObject:
+                    return "Object";
+                case JsonTokenType.StartArray:
+                    return "Array";
+                case JsonTokenType.PropertyName:
+                    return "Property";
+                default:
+                    break;
+            }
+            return tokenType.ToString();
+        }
+
+        public static void EnsureArrayStart(this ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartArray)
+            {
+                throw new InvalidDataException($"Unexpected JSON Token Type '{GetTokenString(reader.TokenType)}'. Expected a JSON Array.");
+            }
+        }
+
+        // Remove after https://github.com/dotnet/corefx/issues/33295 is done
+        public static void Skip(this ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType == JsonTokenType.PropertyName)
+            {
+                reader.Read();
+            }
+
+            if (reader.TokenType == JsonTokenType.StartObject || reader.TokenType == JsonTokenType.StartArray)
+            {
+                int depth = reader.CurrentDepth;
+                while (reader.Read() && depth < reader.CurrentDepth)
+                {
+                }
+            }
+        }
+
+        public static string ReadAsString(this ref Utf8JsonReader reader, byte[] propertyName)
```

Comment:
Similar to ReadAsInt32, if the string can be null, you need to check for TokenType.Null up front first to return null.

This will fail and throw an exception for null values. Poll the TokenType for the null JSON literal, not the ValueSpan.

```C#
        public static int? ReadAsInt32(this ref Utf8JsonReader reader, byte[] propertyName)
        {
            reader.Read();

            if (reader.TokenType == JsonTokenType.Null)
            {
                return null;
            }

            if (reader.TokenType != JsonTokenType.String)
            {
                throw new InvalidDataException($"Expected '{Encoding.UTF8.GetString(propertyName)}' to be of type {JsonTokenType.String}.");
            }

            return reader.GetString();
        }
```

Reply:
I can't think of a situation where we want to accept a null string in the hub protocol.

We explicitly have tests that check for the parsing to fail if null is passed in.

Reply:
Add a `Debug.Assert(reader.TokenType != JsonTokenType.Null)` to be explicit?